src/ott/geometry/regularizers.py

```diff
@@ -0,0 +1,531 @@
+    solver: Linear solver.
+  """
+
+  def __init__(
+      self,
+      A: Optional[lx.TaggedLinearOperator],
+      A_comp: Optional[lx.AbstractLinearOperator] = None,
+      b: Optional[jnp.ndarray] = None,
+      solver: Optional[Callable[[lx.AbstractLinearOperator, jnp.ndarray],
+                                jnp.ndarray]] = None,
+  ):
+    super().__init__()
+    self.A = A
+    self.A_comp = A_comp
+    self.b = b
+    self.solver = (
+        lambda op, b: lx.linear_solve(op, b).value
+    ) if solver is None else solver
+
+  @classmethod
+  def create(
+      cls,
+      A: Optional[Union[jnp.ndarray, lx.AbstractLinearOperator]] = None,
+      *,
+      is_complement: bool = False,
+      is_orthogonal: bool = False,
+      is_squared: bool = False,
+      **kwargs: Any,
+  ) -> "Quadratic":
+    r"""Create the quadratic operator :math:`\frac{1}{2} \left<x, Q x\right> + b`.
+
+    The :math:`Q` is defined as:
+
+    - :math:`Q := A` if not squared and not a complement.
+    - :math:`Q := A^{\perp}` if not squared and a complement.
+    - :math:`Q := A^TA` if squared and not a complement.
+    - :math:`Q := \left(A^{\perp}\right)^TA^{\perp}` if squared and
+      a complement.
+
+    Args:
+      A: Linear operator :math:`A`. If :obj:`None`, use identity.
+      is_complement: Whether to regularize in the orthogonal complement of
+        :math:`A`, defined as :math:`A^{\perp} := I - A^T (AA^T)^{-1} A`.
+      is_orthogonal: Whether :math:`AA^T = I`.
+      is_squared: Whether to square the linear operator.
+      kwargs: Keyword arguments for :class:`Quadratic`.
+    """  # noqa: E501
+    if A is None:
+      return cls(A=None, A_comp=None, **kwargs)
+
+    tags = set()
+    if is_complement:
+      tags.add("complement")
+    if is_squared:
+      tags.add("squared")
+    if is_orthogonal:
+      tags.add("orthogonal")
+
+    if isinstance(A, jnp.ndarray):
+      A = lineax.MatrixLinearOperator(A)
+    A = lx.TaggedLinearOperator(A, tags=tags)
+    A_comp = _complement(A, is_orthogonal) if is_complement else None
+
+    return cls(A, A_comp=A_comp, **kwargs)
+
+  def __call__(self, x: jnp.ndarray) -> float:  # noqa: D102
+    Q = self.Q
+    y = 0.5 * (jnp.dot(x, x) if Q is None else jnp.dot(x, Q.mv(x)))
+    return y if self.b is None else (y + jnp.dot(x, self.b))
+
+  def prox(self, v: jnp.ndarray, tau: float = 1.0) -> jnp.ndarray:  # noqa: D102
+    Q = self.Q
+    b = v if self.b is None else (v - tau * self.b)
+
+    if Q is None:
+      return (1.0 / (1.0 + tau)) * b
+
+    iden = lx.IdentityLinearOperator(Q.out_structure())
+    if self.is_squared:  # use matrix inversion lemma
+      if self.is_complement:
+        # A_comp = I - A^T(AA^T)^{-1}A
+        # prox(v) = (I + tau A_comp^T A_comp)^{-1} (v - tau * b)
+        op = iden + tau * (iden - self.A_comp)
+        return (1.0 / (1.0 + tau)) * op.mv(b)
+      if self.is_orthogonal:
+        op = iden - (tau / (1.0 + tau)) * (self.A.T @ self.A)
+        return op.mv(b)
```

Comment:
What's the reference for this? I couldn't verify this line

Reply:
Ok, let me find it.

Reply:
Found it: https://en.wikipedia.org/wiki/Woodbury_matrix_identity